### ISSUE TITLE
[BUG] Regression on install algorithm since 7.20.2

### ISSUE THREAD

**Issue by Badisi** [npm/cli#3724]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Given the following **package.json** :
```json
{
    "devDependencies": {
        "@hug/eslint-config": "^3.1.2",
        "typescript": "~4.3.5"
    }
}
```
Running `npm install` produces:
```sh
# With npm 7.20.2
node_modules
├── @angular-eslint
│   └── builder
├── @hug
│   └── eslint-config
├── eslint-plugin-rxjs
├── eslint-plugin-rxjs-angular
└── ...

# With npm >= 7.20.3
node_modules
├── @hug
│   └── eslint-config
│       └── node_modules
│           ├── @angular-eslint
│           │   └── builder
│           ├── eslint-plugin-rxjs
│           ├── eslint-plugin-rxjs-angular
│           └── ...
└── ...
```
As explained in the [documentation](https://docs.npmjs.com/cli/v7/commands/npm-install#algorithm), the packages `@angular-eslint`, `eslint-plugin-rxjs`, `eslint-plugin-rxjs-angular` should be installed at the top-level, because they are only required by `@hug/eslint-config` and nowhere else.

This was still the case with `npm@7.20.2` but after that version, packages are now installed privately for `@hug/eslint-config` without even being installed on the top-level.

This is causing me an issue as my angular project is expecting to find `@angular-eslint/builder` relatively from the root node_modules folder of my project.

### Expected Behavior

I expect the install algorithm to work as described in the documentation and as it was working before.

### Environment

- OS: macOS Catalina (version 10.15.6)
- Node: 14.17.5
- npm: 7.22.0

**Comment by ljharb** [comment]
If you’re requiring it from the root, then shouldn’t it be a dep or a dev dep of your project?

**Comment by Badisi** [comment]
I don't need it as a direct dependency.
`@angular-eslint/builder` is a `dependency` of `@hug/eslint-config` which itself is a `devDependency` of my project.
I expect the latter to brings the former to the top-level node_modules of my project.
As shown in my example with `npm@7.20.2`.

The issue I'm getting right now is when `@angular-devkit/architect` is trying to find `@angular-eslint/builder`.
It cannot find it because the module is now deeply installed in a submodule instead of being installed at the top-level as before.
> An unhandled exception occurred: Cannot find module '@angular-eslint/builder/package.json'
> Require stack:
> /src/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js

**Comment by ljharb** [comment]
Right - if `@angular-devkit/architect` is trying to find `@angular-eslint/builder` then the latter needs to be a dependency of the former, otherwise it's relying on dep hoisting, which isn't guaranteed.

**Comment by Badisi** [comment]
Well.. `@angular-devkit/architect` doesn't have a clue about the builder it is supposed to use.
You should install the builder yourself and have it configured in the **angular.json** configuration file of your project.
```json
"lint": {
    "builder": "@angular-eslint/builder:lint"
}
```
But in my case, one of my devDependency (which is responsible of the proper builder version) is already having the builder as a direct dependency, so I found it useless to require it myself too.

-----------

To me the issue is clearly with the npm's install algorithm (and in fact we can see that its behavior changed).
What I don't understand here is why `@angular-eslint/builder` is now installed as a submodule of `@hug/eslint-config`.
The latter is the only one requiring it and I thought that the rule was to install things at the top-level first and as submodules only if they already exists at the top-level under a different version...

**Comment by ljharb** [comment]
I understand that it's implicit for architect - but it should still be an explicit dependency for *you*.

Your scenario is exactly the same as an eslint plugin that's referenced in `.eslintrc` - the plugin still must always be an explicit dev dep so that it's guaranteed to be at the top level of node_modules.

**Comment by Badisi** [comment]
Thanks for your help @ljharb !
I might agree with you and I for sure have a few solutions to solve my problem but that was not really the point of my issue here, so let's put aside the angular and eslint stuff.

What I really want is to understand why it is behaving like this in my specific example.
I mean, `@angular-eslint/builder`, `eslint-plugin-rxjs`, `eslint-plugin-rxjs-angular` are directly installed as submodules and does not even exists on the top-level, how is that possible ?

**Comment by ljharb** [comment]
npm has never guaranteed that things will be hoisted to the top level; it just does that as much as it can to achieve maximal deduping.

It's definitely strange that v7.20.1 and v7.20.2 would have such drastically different node_modules layouts, and hopefully someone on the CLI team can speak to that.

**Comment by Badisi** [comment]
Things changed between `v7.20.2` and `v7.20.3`, but yes I do hope someone can explain what happened.

**Comment by lukekarrys** [comment]
Arborist changed behavior in `2.8.0` which landed in `7.20.3` to fix bugs like https://github.com/npm/cli/issues/3377. I believe the result of that caused the new node_modules shape that you're seeing in later versions. I don't think this is a bug, but I agree that at least the documentation should be updated to reflect the new algorithm.

**Comment by MynockSpit** [comment]
This is either a breaking change or a bug. The current documentation on the dependency algorithm states that dependencies are installed at the top level in cases where there are not multiple versions. If you do not guarantee this, it needs to be explicitly stated.

https://docs.npmjs.com/cli/v8/commands/npm-install#algorithm

---

To further make this weird, the behaviour I'm seeing is dependent on the version syntax that dependents use. 

My `package.json`:
```json
  "dependencies": {
    "first-dep": "^1.0.0"
  }
```

`firstDep`'s `package.json`:
```json
  "dependencies": {
    "dep-a": "1.0.0",
    "dep-b": "^1.0.0"
  }
```

Final tree
```
my-package
├─┬ first-dep@1.0.0
│ └── dep-a@1.0.0
└── deb-b@1.0.0
```

In other words, if `my-package` depends on `some-package` that depends on `dep-a` and it does it at a specific version (and not a `^` version), it won't be hoisted. Otherwise, it will. That doesn't seem like intentional behaviour.
